### PR TITLE
ARCH-251 Script for techdocs push action

### PR DESCRIPTION
**Pull request by glothriel**
Testing this locally is pretty complex, you need to have a correct file and directory structure in place:

* source directory
* destination directory
  * project definition JSON file
* config JSON file

I recommend to play with test scenario (`test_copier.py`) as it tests the whole program e2e, on mock filesystem.

### CODE REVIEW

File: techdocs/script/nodes.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import os
+
+
+# Does not check the file on disk, but uses the path to determine the type
+def looks_fileish(fspath):
+    return "." in os.path.basename(fspath)
+
+def looks_dirish(fspath):
+    return not looks_fileish(fspath)
```
Comment (jzyinq) [inline techdocs/script/nodes.py:9]: That looks suspicioish ;D

why not using `os.path.isfile()` / `os.path.isdir()` ?

`looks_dirish` could have a false positive with hidden directories like `.git`

Some problems with tests? I see that you have such methods after all:
https://github.com/PiwikPRO/actions/blob/3fc4547188eb8b25bd5efc3cc257424c20eada8e/techdocs/script/filesystem.py#L9-L13

File: techdocs/script/config.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+
+@dataclass
+class Config:
+    project: str
+    documents: ConfigDocumentEntry
+
+
+# Validators applied to the whole config
+class RootValidator(abc.ABC):
```
Comment (jzyinq) [inline techdocs/script/config.py:28]:
```suggestion
class ConfigValidator(abc.ABC):
```

File: techdocs/script/config.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    @abc.abstractmethod
+    def validate(self, config: dict):
+        pass
+
+
+# Validators applied to each entry in the documents section
+class EntryValidator(abc.ABC):
```
Comment (jzyinq) [inline techdocs/script/config.py:35]:
```suggestion
class DocumentValidator(abc.ABC):
```

File: techdocs/script/operations.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+                        self.to_path,
+                        relative_dst, 
+                    ),
+                ),
+            )
+        
+
+class DefaultMatcher:
+    def __init__(self, str_to_match):
+        self.regex = re.escape(str_to_match).replace("\\*", ".*")
```
Comment (jzyinq) [inline techdocs/script/operations.py:82]: So for `"some/**/*.txt"` its actually `"some/.*.*/.*.txt"` :D 

Nevertheless glob patterns with such approach are working :tada: 
```json
 "source": "some/**/*.txt",
```
```json
"exclude": [
  "some/directory/with/subdirs/**/*.md"
]
```

**Review by jzyinq — COMMENTED** [review]
Since we have a linter config directory away I suggest to run: 
```bash
black --config ./python/lint/pyproject.toml ./techdocs/script/
```
maybe it's worth a mention in `README.md`/`requirements.txt` as well?

File: techdocs/script/config.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+class DestinationPathMustBePrefixedWithProjectDocsPath:
+    def __init__(self, project_reader: ProjectDetailsReader) -> None:
+        self.reader = project_reader
+
+    def validate(self, project, config):
+        try:
+            if not config["destination"].startswith(self.reader.doc_path(project)):
+                raise ConfigError(
+                    f"Destination path `{config['destination']}` must be prefixed with project docs path `{self.reader.doc_path(project)}`"
+                )
```
[the remark this answers is not in the document]
Comment (glothriel) [reply techdocs/script/config.py:177]: I also though about this, but it creates an edge case I don't like:

* given project directory `promil`
* If for whatever crazy reason I'd like to put a file into `promil/promil/file.md` the `prepender` will interpret the path as "Ah, the prefix is already there". One way to go around will be to put the absolute path including the prefix as a destination, but I decided to give it up in favor of less forgivable but clearer approach.

File: techdocs/script/README.md [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+Example structure of a config file:
+
+```json
+{
+    "project": "promil",
+    "documents": [
+        {"source": "README.md", "destination": "docs/promil/bla.md"},
+        {"source": "docs/*", "destination": "docs/promil/somedir", "exclude": ["docs/internal/*"]},
+    ],
+}
```
[the remark this answers is not in the document]
Comment (glothriel) [reply techdocs/script/README.md:20]: Then you call the github action multiple times, imo there's no reason to complicate this.

File: techdocs/script/config.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jzyinq) [reply techdocs/script/config.py:177]: After offline discussion we agreed to remove that error and prepend project path to `destination`.

File: techdocs/script/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (jzyinq) [reply techdocs/script/README.md:20]: As we discussed offline handling `push-docs-per-projects` implies problems with removing unwanted files in current tech docs structure and we decided to handle mutliple project in single configuration: 

```json
{
  "documents": [
    {
      "project": "promil",
      "source": "docs/*",
      "destination": "stacks/Promil-stack-analytics",
      "exclude": [
        "docs/README.md",
        "docs/internal/*"
      ]
    },
    {
      "project": "other-project",
      "source": "docs/README.md",
      "destination": "analytics-stack.md"
    }
  ]
}
```

**Review by jzyinq — APPROVED** [review]

**Pull request merged**